elife/multi-process-test.sls

```diff
@@ -0,0 +1,48 @@
+
+# test that N processes are brought using a single service state
+# each process simply writes it's identifier and the time to the syslog
+
+{% set process = "test" %}
+{% set num_processes = 10 %}
+
+/opt/{{ process }}.py:
+    file.managed:
+        - source: salt://elife/scripts/test.py
+
+{{ process }}:
+    file.managed:
+        - name: /lib/systemd/system/{{ process }}@.service
+        - source: salt://elife/config/lib-systemd-system-test.service
+        - template: jinja
+        - context:
+            process: {{ process }}
+        - require:
+            - /opt/{{ process }}.py
+
+    # neccessary because salt mangles "service@{x..y}" to "service@\x7bx..y\x7d.service" 
+    # systemd will see that as just a single instantiated service
+    cmd.run:
+        - name: |
+            set -e
+            systemctl disable {{ process }}@{0..99} # disable all instances (up to 99). implicit reload
+            systemctl enable {{ process }}@{0..1} # enable just the range we're after. implicit reload
+        - require:
+            - file: {{ process }}
+        # only run command if there are disabled instances in our range
+        # todo: only runs if range grows. if range shrinks to 0..0 from 0..1 then this command won't be triggered
+        - onlyif:
+            - systemctl is-enabled {{ process }}@{0..1} | grep -q 'disabled'
+
+{{ process }}-controller.target:
+    file.managed:
+        - name: /lib/systemd/system/{{ process }}-controller.target
+        - source: salt://elife/templates/process-controller.target
+
+    service.running:
```

Comment:
This state does not restart the processes:
```
          ID: test-controller.target
    Function: service.running
      Result: True
     Comment: The service test-controller.target is already running
     Started: 10:21:43.602231
    Duration: 30.64 ms
     Changes:  
```
but
```
sudo systemctl restart test-controller.target
```
does. Include that with a `cmd.run`?

Reply:
it's not going to restart the state if it's already running, no.

you would want to restart the state if the number of processes running is different from what you intended, whether one has died or more added/removed.

this target/unit/state/service thing doesn't care if the processes that depend on it are there or not, so it's probably not ever going to fail in coming up or being restarted.